src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -747,9 +766,33 @@
   should be copied to."
   ([cacrl :- schema/Str
     localcacrl :- schema/Str]
-    (when (fs/exists? cacrl)
-      (ks/mkdirs! (fs/parent localcacrl))
-      (fs/copy cacrl localcacrl))))
+   (when (not= cacrl localcacrl)
+     (let [max-attempts 5]
+       (when (and (fs/exists? cacrl))
+         (ks/mkdirs! (fs/parent localcacrl))
+         ;; Make up to 'max-attempts' tries to copy the cacrl file to the
+         ;; localcacrl file.  The content of the cacrl file is read into memory
+         ;; and parsed for valid CRL pem content during each attempt.  The content
+         ;; in memory is written to the localcacrl file only if it is valid.  This
+         ;; validation is done to protect against a partially written (which could
+         ;; happen if an asynchronous revocation is in process) or corrupt
+         ;; cacrl file being copied.
+         (loop [attempts-left max-attempts
+                ca-crl-last-modified (fs/mod-time cacrl)]
+           (let [cacrl-as-string (slurp cacrl)]
+             (when-let [write-exception (write-local-cacrl!
+                                         localcacrl
+                                         cacrl-as-string)]
+               (if (zero? attempts-left)
+                 (log/error (i18n/trs "Unable to synchronize crl file {0} to {1}: {2}"
+                                      cacrl localcacrl (.getMessage write-exception)))
```

Comment:
Can we add some more context to this error message so folks know what the implications are and what to try next? I'm thinking something like:
...
```
Unable to synchronize crl file {0} to {1}: {2}. Recent changes to the 
CRL may not have taken effect. To load the updated CRL file reload 
or restart the Puppet Server service."
```
 Or something like that.